language/move-compiler/src/naming/translate.rs

```diff
@@ -495,10 +543,15 @@ fn function(
         body,
         specs: _,
     } = ef;
+    assert!(context.local_scopes.is_empty());
+    assert!(context.local_count.is_empty());
+    assert!(context.used_locals.is_empty());
+    context.local_scopes = vec![BTreeMap::new()];
+    context.local_count = BTreeMap::new();
```

Comment:
Is there a need to partition this count by the name, or would it be okay to just count the number of locals in the function?  I.e. instead of this:

```
fun foo(): u64 {
  let x = 39;
  let y = x + 1;
  let x = y + 1;
  let y = x + 1;
  y
}
```

Lowering to:

```
fun foo(): u64 {
  let x#0#0 = 39;
  let y#0#0 = x#0#0 + 1;
  let x#1#0 = y#0#0 + 1;
  let y#1#0 = x#1#0 + 1;
  y#1#0
}
```

It becomes:

```
fun foo(): u64 {
  let x#0#0 = 39;
  let y#1#0 = x#0#0 + 1;
  let x#2#0 = y#1#0 + 1;
  let y#3#0 = x#2#0 + 1;
  y#3#0
}
```

(I think you could also potentially simplify away `color` using this approach by bumping all the vars in a macro expansion or inline by the number of locals you've used so far in the function body)

Reply:
I was wondering if there was some value in being able to tell which instance of a given name it is. But that probably isn't true. 

The color thing is still really helpful (though not strictly necessary) when expanding macros as it lets you swap between "scopes" a bit easier, which is something needed when expanding the blocks/lambdas in macros.

Reply:
Actually, just went to change this. I might leave it as is? Even though it isn't necessary, I rely on the count being `0` for parameters for ever so slightly better error messages. 

Though I could still set the count to 0 for parameters.

Interestingly though, if I do just user the number and increment it each time, I don't even really need the name except for display purposes. I'm not sure that's helpful anywhere, but it is interesting.

Reply:
Noted on `color` -- I think it also becomes helpful if/when macros are able to define new macros (vague recollection on my part, I may be wrong).

Re: detecting parameters, you could do this by making sure the first `p` indices in a p-arity function are reserved for its parameters. This does break down if you are relying on this property during inlining though (you won't necessarily know how many parameters there are in the function the body was inlined from).

The aim here was to simplify things slightly, so not a dealbreaker to leave it.